blockchain/chain.go

```diff
@@ -503,16 +503,37 @@ func (b *BlockChain) getReorganizeNodes(node *blockNode) (*list.List, *list.List
 	if node == nil {
 		return detachNodes, attachNodes
 	}
+	if node.parent.KnownInvalid() {
+		node.status |= statusInvalidAncestor
+		return detachNodes, attachNodes
+	}
 
 	// Find the fork point (if any) adding each block to the list of nodes
 	// to attach to the main tree.  Push them onto the list in reverse order
 	// so they are attached in the appropriate order when iterating the list
 	// later.
 	forkNode := b.bestChain.FindFork(node)
+	invalidChain := false
 	for n := node; n != nil && n != forkNode; n = n.parent {
+		if n.KnownInvalid() {
+			invalidChain = true
+			break
+		}
 		attachNodes.PushFront(n)
 	}
 
+	// If any of the node's ancestors are invalid, unwind attachNodes, marking
+	// each one as invalid for future reference.
+	if invalidChain {
+		var next *list.Element
+		for e := attachNodes.Front(); e != nil; e = next {
+			next = e.Next()
+			n := attachNodes.Remove(e).(*blockNode)
+			n.status |= statusInvalidAncestor
```

Comment:
What do you think about having a `func (node *blockNode) SetStatus(status blockStatus)` that accepts the status flag(s)?  My reasoning is that when this is persisted, chances are there will need to be some type of map to track the modified nodes for a later database flush.  This will be particularly important with a utxocache since the states have to be kept in sync in memory.  Having a setter means it would only need to marked modified once in the setter versus having to remember to update the modified map after setting it all over the place.

Reply:
As you can see in https://github.com/btcsuite/btcd/pull/1066, I decided to to explicitly call `MarkDirty` where required. This was because there were times (in initChainState) when we would want to add nodes without marking the blocks dirty. I guess that could be handled differently though, by clearing the dirty set or somehow having it inactive until `initChainState` completes.

Would `SetStatus` directly replace the status or do the bitwise or? Most of the time we edit the status we want to do the bitwise or, so I think `SetStatusFlags` and `UnsetStatusFlags` would be more convenient.